tests/correctness/sync/sync-dot-product-opt.futil

```diff
@@ -0,0 +1,257 @@
+      wait_compute[done] = flag_reg.done;
+    }
+
+    comb group has_computing_job {
+      eq.left = flag.out;
+      eq.right = 1'd1;
+    }
+
+    group initialize_idx {
+      idx_0.in = in_0.read_data;
+      in_0.addr0 = 32'd0;
+      idx_0.write_en = 1'd1;
+      idx_1.in = in_1.read_data;
+      in_1.addr0 = 32'd0;
+      idx_1.write_en = 1'd1;
+      initialize_idx[done] = idx_0.done & idx_1.done? 1'd1;
+    }
+
+    group initialize_val {
+      val_0.in = in_0.read_data;
+      in_0.addr0 = 32'd1;
+      val_0.write_en = 1'd1;
+      val_1.in = in_1.read_data;
+      in_1.addr0 = 32'd1;
+      val_1.write_en = 1'd1;
+      initialize_val[done] = val_0.done & val_1.done? 1'd1;
+    }
+
+    group fwd_idx_0 {
+      idx_0.in = in_0.read_data;
+      in_0.addr0 = point_0.out;
+      idx_0.write_en = 1'd1;
+      fwd_idx_0[done] = idx_0.done; 
+    }
+
+    group fwd_idx_1 {
+      idx_1.in = in_1.read_data;
+      in_1.addr0 = point_1.out;
+      idx_1.write_en = 1'd1;
+      fwd_idx_1[done] = idx_1.done; 
+    }
+
+    group forward_pointer_0 {
+      incr_0.left =32'd2;
+      incr_0.right = point_0.out;
+      point_0.in = incr_0.out;
+      point_0.write_en = 1'd1;
+      forward_pointer_0[done] = point_0.done;
+    }
+
+    group forward_pointer_1 {
+      incr_1.left =32'd2;
+      incr_1.right = point_1.out;
+      point_1.in = incr_1.out;
+      point_1.write_en = 1'd1;
+      forward_pointer_1[done] = point_1.done;
+    }
+
+    group val_to_reg_0 {
+      incr_0.left = 32'd1;
+      incr_0.right = point_0.out;
+      val_0.in = in_0.read_data;
+      in_0.addr0 = incr_0.out;
+      val_0.write_en = 1'd1;
+      val_to_reg_0[done] = val_0.done;
+    }
+
+    group val_to_reg_1 {
+      incr_1.left = 32'd1;
+      incr_1.right = point_1.out;
+      val_1.in = in_1.read_data;
+      in_1.addr0 = incr_1.out;
+      val_1.write_en =1'd1;
+      val_to_reg_1[done] = val_1.done;
+    }
+
+    group compute_product {
+      mult.go = 1'd1;
+      mult.left = val_0.out;
+      mult.right = val_1.out;
+      val_out.in = mult.done? mult.out;
+      val_out.write_en = mult.done?1'd1;
+      compute_product[done] = val_out.done;
+    }
+
+    group add_to_out {
+      add.left = out.read_data;
+      out.addr0 = 32'd0;
+      add.right = val_out.out;
+      out.write_data = add.out;
+      out.write_en = 1'd1;
+      add_to_out[done] = out.done;
+    }
+
+    comb group chase_0 {
+      lt_0.left = idx_0.out;
+      lt_0.right = idx_1.out;
+    }
+
+    comb group chase_1 {
+      lt_1.left = idx_1.out;
+      lt_1.right = idx_0.out;
+    }
+
+    comb group comp {
+      fwd_0.left = idx_0.out;
+      fwd_0.right = 32'd16;
+      fwd_1.left = idx_1.out;
+      fwd_1.right = 32'd16;
+      signal.left = fwd_0.out;
+      signal.right = fwd_1.out;
+    }
+
+  }
+
+  control {
+    seq {
+      initialize_idx;
+      initialize_val;
+      par {
+        // thread A
+        while signal.out with comp {
```

Comment:
Can you add some documentation on the structure of the threads and the high-level idea behind this architecture? It would be difficult to debug this design in the future without something like that